pkg/controller/nodeobservatory/controller.go

```diff
@@ -0,0 +1,379 @@
+		AddFunc:    controller.klusterAddFunc,
+		UpdateFunc: controller.klusterUpdateFunc,
+		DeleteFunc: controller.klusterDeleteFunc,
+	})
+
+	return controller
+}
+
+func (n *NodeObservatory) Run(threadiness int, stopCh <-chan struct{}, wg *sync.WaitGroup) {
+	n.logger.Log(
+		"msg", "starting run loop",
+		"threadiness", threadiness,
+		"v", 2,
+	)
+
+	n.stopCh = stopCh
+	n.informerWg = wg
+
+	defer n.queue.ShutDown()
+	defer wg.Done()
+	wg.Add(1)
+
+	for i := 0; i < threadiness; i++ {
+		go wait.Until(n.runWorker, time.Second, stopCh)
+	}
+
+	ticker := time.NewTicker(KlusterRecheckInterval)
+	go func() {
+		for {
+			select {
+			case <-ticker.C:
+				n.requeueAllKlusters()
+			case <-stopCh:
+				ticker.Stop()
+				return
+			}
+		}
+	}()
+	<-stopCh
+}
+
+func (n *NodeObservatory) requeueAllKlusters() (err error) {
+	defer func() {
+		n.logger.Log(
+			"msg", "requeued all",
+			"v", 1,
+			"err", err,
+		)
+	}()
+
+	klusters, err := n.Factories.Kubernikus.Kubernikus().V1().Klusters().Lister().List(labels.Everything())
+	if err != nil {
+		return err
+	}
+
+	for _, kluster := range klusters {
+		n.requeueKluster(kluster)
+	}
+
+	return nil
+}
+
+func (n *NodeObservatory) requeueKluster(kluster *v1.Kluster) {
+	n.logger.Log(
+		"msg", "queuing",
+		"kluster", kluster.Spec.Name,
+		"project", kluster.Account(),
+		"v", 2,
+	)
+
+	key, err := cache.MetaNamespaceKeyFunc(kluster)
+	if err == nil {
+		n.queue.Add(key)
+	}
+}
+
+func (n *NodeObservatory) runWorker() {
+	for n.processNextWorkItem() {
+	}
+}
+
+func (n *NodeObservatory) processNextWorkItem() bool {
+	key, quit := n.queue.Get()
+	if quit {
+		return false
+	}
+	defer n.queue.Done(key)
+
+	var kluster *v1.Kluster
+	var requeue bool
+
+	if obj, exists, _ := n.Factories.Kubernikus.Kubernikus().V1().Klusters().Informer().GetIndexer().GetByKey(key.(string)); exists {
+		kluster = obj.(*v1.Kluster)
+	}
+
+	// Invoke the method containing the business logic
+	requeue, err := n.reconcile(kluster)
+
+	if err == nil {
+		// Forget about the #AddRateLimited history of the key on every successful synchronization.
+		// This ensures that future processing of updates for this key is not delayed because of
+		// an outdated error history.
+
+		if requeue == false {
+			n.queue.Forget(key)
+		} else {
+			// Requeue requested
+			n.queue.AddAfter(key, BaseDelay)
+		}
+
+		return true
+	}
+
+	if n.queue.NumRequeues(key) < 5 {
+		// Re-enqueue the key rate limited. Based on the rate limiter on the
+		// queue and the re-enqueue history, the key will be processed later again.
+		n.queue.AddRateLimited(key)
+		return true
+	}
+
+	// Retries exceeded. Forgetting for this reconciliation loop
+	n.queue.Forget(key)
+	return true
+}
+
+func (n *NodeObservatory) reconcile(kluster *v1.Kluster) (bool, error) {
+
+	n.cleanUpInformers()
+
+	if err := n.createAndWatchNodeInformerForKluster(kluster); err != nil {
```

Comment:
We should exit before this if if kluster is `nil`. Pretty sure we get a panic otherwise.

Reply:
true. fixed